### ISSUE TITLE
Add minfreespace flag for creation policies from mergerfs 

### ISSUE THREAD

**Issue by joshuakohut77** [rclone/rclone#6071]
#### The associated forum post URL from `https://forum.rclone.org`
https://forum.rclone.org/t/union-remote-doesn-t-switch-when-first-remote-is-full-update

#### What is your current rclone version (output from `rclone version`)?
```
rclone v1.58.0
os/version: ubuntu 20.04 (64 bit)
os/kernel: 5.11.0-41-generic (x86_64)
os/type: linux
os/arch: amd64
go/version: go1.17.8
go/linking: static
go/tags: none
```

#### What problem are you are trying to solve?
rclone being unable to successfully determine how much free space is on a drive

#### How do you think rclone should be changed to solve that?
Add `minfreespace` flag for creation policies from mergerfs.
minfreespace=SIZE: The minimum space value used for creation policies. Can be overridden by branch specific option. Understands 'K', 'M', and 'G' to represent kilobyte, megabyte, and gigabyte respectively. (default: 4G)
If available disk space is below the minfreespace, move to a new disk. 

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by ncw** [comment]
The problem with using one of minimum free space policies (`eplfs` or `lfs`) is they don't stop uploading until there is no space left at all.

This means that if you restart the union after one remote is full, rclone will continue to try to fill it up.

It is hard in the general case to work out if a file will fit or not on a remote (we may not know the size - this is all files with `rclone mount` and `--vfs-cache-mode off` for example).

A workaround for this is to not fill remotes beyond `minfreespace` free space.

So instead of finding the remote with the minimum free space, we find the remote with the minimum free space, provided it is above `minfreespace`.

The docs currently say

eplfs (existing path, least free space) | Of all the upstreams on which the relative path exists choose the one with the least free space.
-- | --

So we'd be changing that to

eplfs (existing path, least free space) | Of all the upstreams on which the relative path exists choose the one with the least free space provided it is bigger than the `minfreespace` parameter.
-- | --

Would you agree?

@joshuakohut77 do you see a single configurable parameter for this? So you'd set `--union-min-free-space 1G` say and it would apply to all remotes? I'm not sure exactly how mergerfs does it here (and we don't need to do it exactly the same way if we don't want to!).

**Comment by joshuakohut77** [comment]
A single configurable parameter such as `--union-min-free-space 1G` would be the most simple version to accomplish at least my use case. I am not familiar with how either rclone or mergerfs works on the backend so I'll add some ideas here that may not make sense or not feasible. 
From reading other challenging things about storing on remotes, determining if they're full or when they're going to be full is a difficult task. Another way to accomplish the same thing would be to have a property of the remote during config. That property would be a `min-free-space` threshold. This would allow you to configure each remote rather than the union. It would be the same comparison to check the remotes' free space to it's threshold. 
Having extra configurability is always nice but I can't think of a use case where each drive being configured vs one parameter for all remotes adds any real benefit. If someone is using a union, they are intending on treating it as one big drive and therefore probably don't care about each individual drive. 

Obviously this `min-free-space` parameter is only useful as long as the value is larger than your largest file you are storing. From the little bit I understand, the issue only occurs during the initial write to the remote because after that the existing eplfs rule of writing until it's full will take over. 

So in summary, I agree that a single configurable parameter to apply to all remotes in a union would suffice. This would be a great improvement for anyone who's storing small files like documents or pictures.

**Comment by ncw** [comment]
> A single configurable parameter such as `--union-min-free-space 1G` would be the most simple version to accomplish at least my use case.

I agree. It is possible we might make something more sophisticated in future, but that should do for now.

I started writing instructions for you to follow to implement, then I realised it is a lot more complicated than I first thought since the options aren't being passed to the policies.

So I've implemented that here. I've also fixed a number of bugs along the way in the lfs policy.

[v1.59.0-beta.6055.7a26a9746.fix-6071-union](https://beta.rclone.org/branch/fix-6071-union/v1.59.0-beta.6055.7a26a9746.fix-6071-union/) on branch [fix-6071-union](https://github.com/rclone/rclone/tree/fix-6071-union) (uploaded in 15-30 mins)

Just to confirm - you are using lfs/eplfs policy aren't you? In the forum your config didn't have the policy set and the default is epmfs...

**Comment by joshuakohut77** [comment]
You are absolutely correct that I didn't have lfs/eplfs when I posted that config. That was a result of trying different policies in an attempt to find one that works for what I need. 
However I installed the beta branch you linked, I tested the new parameter using lfs policy and it's worked exactly as expected. I first ran it without the new parameter. It filled up one of the 15G drives and then errored about the storage drive being full. 
I added the parameter `--union-min-free-space 50M` because all my files are 50M or smaller and it worked perfectly. I will be running this sync every night on a cron job so I will let you know if I run into any errors. 

Thank you very much for your help on this. It added a lot of value with how I wanted to use my union drive.

**Comment by ncw** [comment]
Great - glad that is working.

I set the default for that parameter, somewhat arbitrarily to 1G. Do you have a feeling as to what it should be? I felt 1G was probably large enough to keep most people out of trouble but small enough so that it doesn't waste a lot of space.

**Comment by joshuakohut77** [comment]
1G sounds fine to me. You're always going to have someone who disagrees. All you can do is try and do the best for the majority of people. If someone finds this thread in the future and has a compelling argument why it should be different they can always comment. I believe this was a great success. I will let you close out when you're ready. 
Thank you again for all your help!
